src/registry/otlp/mod.rs

```diff
@@ -0,0 +1,607 @@
+            StopSignal::AdminStop => f.write_str("ADMIN_STOP"),
+            StopSignal::Inactivity => f.write_str("INACTIVITY"),
+        }
+    }
+}
+
+/// Start an OTLP receiver listening to a specific port on all IPv4 interfaces
+/// and return an iterator of received OTLP requests.
+pub fn listen_otlp_requests(
+    grpc_port: u16,
+    admin_port: u16,
+    inactivity_timeout: Duration,
+    logger: impl Logger + Sync + Clone,
+) -> impl Iterator<Item = OtlpRequest> {
+    let addr = format!("0.0.0.0:{grpc_port}")
+        .parse()
+        .expect("Failed to parse address");
+    let (tx, rx) = mpsc::channel(100);
+    let stop_tx = tx.clone();
+    // Create a watch channel for the last activity timestamp
+    let (activity_tx, activity_rx) = watch::channel(Instant::now());
+    let logs_service = LogsServiceImpl {
+        tx: tx.clone(),
+        activity_tx: activity_tx.clone(),
+    };
+    let metrics_service = MetricsServiceImpl {
+        tx: tx.clone(),
+        activity_tx: activity_tx.clone(),
+    };
+    let trace_service = TraceServiceImpl {
+        tx: tx.clone(),
+        activity_tx: activity_tx.clone(),
+    };
+
+    logger.log("To stop the OTLP receiver:");
+    logger.log("  - press CTRL+C,");
+    logger.log(&format!(
+        "  - send a SIGHUP signal to the weaver process or run this command kill -SIGHUP {}",
+        std::process::id()
+    ));
+    logger.log(&format!("  - or send a POST request to the /stop endpoint via the following command curl -X POST http://localhost:{}/stop.", grpc_port + 1));
+    logger.log(&format!(
+        "The OTLP receiver will stop after {} seconds of inactivity.",
+        inactivity_timeout.as_secs()
+    ));
+
+    // Start an OS thread and run a single threaded Tokio runtime inside.
+    // The async OTLP receiver sends the received OTLP messages to the Tokio channel.
+    let _ = std::thread::spawn(move || {
+        // Start a current threaded Tokio runtime
+        tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("Failed to build Tokio Runtime")
+            .block_on(async {
+                let mut tasks = JoinSet::new();
+
+                // Spawn tasks to handle different stop signals
+                spawn_stop_signal_handlers(stop_tx.clone(), &mut tasks);
+                spawn_http_stop_handler(stop_tx.clone(), admin_port, &mut tasks).await;
+                spawn_inactivity_monitor(
+                    stop_tx.clone(),
+                    activity_rx,
+                    inactivity_timeout,
+                    &mut tasks,
+                );
+
+                // Serve the OTLP services
+                let server = Server::builder()
+                    .add_service(LogsServiceServer::new(logs_service))
+                    .add_service(MetricsServiceServer::new(metrics_service))
+                    .add_service(TraceServiceServer::new(trace_service))
+                    .serve(addr)
```

Comment:
If something is already listening on `addr` there should be bind error here. I'm not getting this in my testing - it just silently fails to bind.

Reply:
Missed this one. I will fix that.

Reply:
After taking a closer look at the code, I do have error handling in the statement that follows. This statement involves sending the error to the channel, which is exposed as an iterator on the consumer side. The fact that you don’t see an error message is probably due to Weaver (or the test) stopping before the consumer has had time to process the error. 

I’ll fix that.

Reply:
Fixed.

Reply:
That doesn't actually fix my problem, but I know why. I didn't spot that you're binding on `0.0.0.0:4317` I have another process bound on `127.0.0.1:4317`. So the second process is successfully binding on other interfaces, just not loopback. Then I send to `127.0.0.1:4317` and of course don't receive. Perhaps the fix here is to allow the user to enter the whole address not just the port - that's probably better for security anyway.